### ISSUE TITLE
Premature end of file

### ISSUE THREAD

**Issue by ktk** [zazuko/carml-service#6]
Version 0.0.3 does not work reliable, at least in some of the settings we have. In GitLab pipelines it very quickly fails with this error:

```
info:   CARML_SERVICE: http://carml-service:8080/ {"iri":"http://example.org/pipeline/toFile","timestamp":"2023-08-11T14:07:58.037Z"}
info: created new Step {"iri":"http://example.org/pipeline/parseN3","timestamp":"2023-08-11T14:07:58.041Z"}
Error: server error: [400] Bad Request ({"title":"Error executing XPath expression.","detail":"io.carml.logicalsourceresolver.LogicalSourceResolverException: Error executing XPath expression.\n\tat io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux(XPathResolver.java:154)\n\tat io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2(XPathResolver.java:120)\n\tat reactor.core.publisher.FluxCreate.subscribe(FluxCreate.java:95)\n\tat reactor.core.publisher.Flux.subscribe(Flux.java:8660)\n\tat reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:200)\n\tat reactor.core.publisher.FluxFlatMap.subscribeOrReturn(FluxFlatMap.java:93)\n\tat reactor.core.publisher.Flux.subscribe(Flux.java:8646)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426)\n\tat reactor.core.publisher.FluxIterable$IterableSubscription.slowPath(FluxIterable.java:335)\n\tat reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:294)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:371)\n\tat reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:201)\n\tat reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:83)\n\tat reactor.core.publisher.Flux.subscribe(Flux.java:8660)\n\tat reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:258)\n\tat reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:78)\n\tat reactor.core.publisher.Mono.subscribe(Mono.java:4444)\n\tat reactor.core.publisher.Mono.block(Mono.java:1733)\n\tat io.carml.engine.rdf.RdfRmlMapper.toModel(RdfRmlMapper.java:368)\n\tat io.carml.engine.rdf.RdfRmlMapper.mapToModel(RdfRmlMapper.java:351)\n\tat com.zazuko.service.carml.CarmlEndpoint.doConversion(CarmlEndpoint.java:68)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.directProceed(AbstractInvocationContext.java:113)\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.proceed(AbstractInvocationContext.java:106)\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:78)\n\tat org.apache.meecrowave.cxf.JAXRSFieldInjectionInterceptor.lazyInjectContexts(JAXRSFieldInjectionInterceptor.java:64)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat org.apache.webbeans.component.InterceptorBean.intercept(InterceptorBean.java:136)\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:65)\n\tat org.apache.webbeans.intercept.DefaultInterceptorHandler.invoke(DefaultInterceptorHandler.java:139)\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbInterceptProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbNormalScopeProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)\n\tat org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)\n\tat org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)\n\tat org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)\n\tat org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265)\n\tat org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)\n\tat org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:681)\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)\n\tat org.apache.meecrowave.cxf.CxfCdiAutoSetup$1.doFilter(CxfCdiAutoSetup.java:122)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)\n\tat org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)\n\tat org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382)\n\tat org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)\n\tat org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732)\n\tat org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n\tSuppressed: java.lang.Exception: #block terminated with an error\n\t\tat reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:139)\n\t\tat reactor.core.publisher.Mono.block(Mono.java:1734)\n\t\t... 61 more\nCaused by: javax.xml.xpath.XPathException: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.\njavax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.\n\tat jlibs.xml.sax.dog.XMLDog.sniff(XMLDog.java:192)\n\tat io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux(XPathResolver.java:152)\n\t... 78 more\nCaused by: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.\njavax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.\n\tat io.carml.logicalsourceresolver.PausableStaxXmlReader.run(PausableStaxXmlReader.java:152)\n\tat io.carml.logicalsourceresolver.PausableStaxXmlReader.start(PausableStaxXmlReader.java:103)\n\tat io.carml.logicalsourceresolver.PausableStaxXmlReader.parse(PausableStaxXmlReader.java:75)\n\tat jlibs.xml.sax.dog.XMLDog.sniff(XMLDog.java:189)\n\t... 79 more\nCaused by: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.\n\tat java.xml/com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next(XMLStreamReaderImpl.java:652)\n\tat io.carml.logicalsourceresolver.PausableStaxXmlReader.run(PausableStaxXmlReader.java:150)\n\t... 82 more\n"})
    at Client.map (/builds/pipelines/fso-ech_0071-ld/node_modules/barnard59-carml-service/lib/Client.js:39:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /builds/pipelines/fso-ech_0071-ld/node_modules/barnard59-carml-service/transform.js:17:26
```

So the error seems to be `Message: Premature end of file.`. This can be "fixed" by trying often enough but at least on the ldbar Gitlab it fails more often than not. I wonder if it is related to amount of memory/cpu assigned, as in that it fails less on beefy machines but more on restricted machines.

The log above is taken from [this run](https://gitlab.ldbar.ch/pipelines/fso-ech_0071-ld/-/jobs/63822), I had to run it many times to get one done without errors. 

I propose to update to the latest carml release & if necessary add additional debugging support so that we can narrow down how/why the issue happens.

**Comment by mchlrch** [comment]
I updated the dependencies to the latest carml version (0.4.8)

I can reproduce the `javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Premature end of file.` if I call the service with an empty XML source attachment.
 
On the other hand, calling the service with an empty JSON source attachment doesn't fail, but succeeds with HTTP 200 (Content-Length: 0). That's seems reasonable. Now I modified the code to setup and run carml only if the source attachment is not blank. That way, empty XML source attachments behave the same as empty JSON source attachments.

**Comment by mchlrch** [comment]
This is how it behaves now after the modifications:
```
$ curl -i -F mapping=@mapping.carml.ttl -F source=@example.json -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 100 

HTTP/1.1 200 
Date: Wed, 30 Aug 2023 19:53:43 GMT
Content-Type: text/turtle
Content-Length: 350


<http://example.org/character/0> a <http://schema.org/Person>;
  <http://example.org/hairColor> "black";
  <http://schema.org/lastName> "JsonSmith";
  <http://schema.org/givenName> "Bob" .

<http://example.org/character/1> a <http://schema.org/Person>;
  <http://example.org/hairColor> "orange";
  <http://schema.org/givenName> "Charlie" .

$ curl -i -F mapping=@mapping.carml-xpath.ttl -F source=@example.xml -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 100 

HTTP/1.1 200 
Date: Wed, 30 Aug 2023 19:53:54 GMT
Content-Type: text/turtle
Content-Length: 349


<http://example.org/character/0> a <http://schema.org/Person>;
  <http://schema.org/givenName> "Bob";
  <http://example.org/hairColor> "black";
  <http://schema.org/lastName> "XmlSmith" .

<http://example.org/character/1> a <http://schema.org/Person>;
  <http://schema.org/givenName> "Charlie";
  <http://example.org/hairColor> "orange" .

$ curl -i -F mapping=@mapping.carml.ttl -F source=@empty-source.json -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 100 

HTTP/1.1 200 
Date: Wed, 30 Aug 2023 19:54:08 GMT
Content-Type: text/turtle
Content-Length: 0

$ curl -i -F mapping=@mapping.carml-xpath.ttl -F source=@empty-source.xml -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 100 

HTTP/1.1 200 
Date: Wed, 30 Aug 2023 19:55:14 GMT
Content-Type: text/turtle
Content-Length: 0

$ curl -i -F mapping=@empty-mapping.ttl -F source=@example.json -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 400 
Date: Wed, 30 Aug 2023 19:55:45 GMT
Content-Type: application/problem+json
Content-Length: 5741
Connection: close

{"title":"No actionable triples maps provided.","detail":"io.carml.engine.RmlMapperException: No actionable triples maps provided.\r\n\tat io.carml.engine.rdf.RdfRmlMapper$Builder.build(RdfRmlMapper.java:222)\r\n\tat com.zazuko.service.carml.CarmlEndpoint.doConversion(CarmlEndpoint.java:69)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.directProceed(AbstractInvocationContext.java:113)\r\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.proceed(AbstractInvocationContext.java:106)\r\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:78)\r\n\tat org.apache.meecrowave.cxf.JAXRSFieldInjectionInterceptor.lazyInjectContexts(JAXRSFieldInjectionInterceptor.java:64)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.webbeans.component.InterceptorBean.intercept(InterceptorBean.java:136)\r\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:65)\r\n\tat org.apache.webbeans.intercept.DefaultInterceptorHandler.invoke(DefaultInterceptorHandler.java:139)\r\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbInterceptProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\r\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbNormalScopeProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)\r\n\tat org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)\r\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)\r\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)\r\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)\r\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)\r\n\tat org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)\r\n\tat org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)\r\n\tat org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)\r\n\tat org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)\r\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:681)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)\r\n\tat org.apache.meecrowave.cxf.CxfCdiAutoSetup$1.doFilter(CxfCdiAutoSetup.java:122)\r\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)\r\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)\r\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)\r\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)\r\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540)\r\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)\r\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)\r\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)\r\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)\r\n\tat org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382)\r\n\tat org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)\r\n\tat org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895)\r\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732)\r\n\tat org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)\r\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)\r\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)\r\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\r\n\tat java.base/java.lang.Thread.run(Thread.java:834)\r\n"}

$ curl -i -F mapping=@empty-mapping.ttl -F source=@example.xml -H "Accept: text/turtle" http://mauna.local:8080/
HTTP/1.1 400 
Date: Wed, 30 Aug 2023 19:56:00 GMT
Content-Type: application/problem+json
Content-Length: 5741
Connection: close

{"title":"No actionable triples maps provided.","detail":"io.carml.engine.RmlMapperException: No actionable triples maps provided.\r\n\tat io.carml.engine.rdf.RdfRmlMapper$Builder.build(RdfRmlMapper.java:222)\r\n\tat com.zazuko.service.carml.CarmlEndpoint.doConversion(CarmlEndpoint.java:69)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.directProceed(AbstractInvocationContext.java:113)\r\n\tat org.apache.webbeans.intercept.AbstractInvocationContext.proceed(AbstractInvocationContext.java:106)\r\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:78)\r\n\tat org.apache.meecrowave.cxf.JAXRSFieldInjectionInterceptor.lazyInjectContexts(JAXRSFieldInjectionInterceptor.java:64)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.webbeans.component.InterceptorBean.intercept(InterceptorBean.java:136)\r\n\tat org.apache.webbeans.intercept.InterceptorInvocationContext.proceed(InterceptorInvocationContext.java:65)\r\n\tat org.apache.webbeans.intercept.DefaultInterceptorHandler.invoke(DefaultInterceptorHandler.java:139)\r\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbInterceptProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\r\n\tat com.zazuko.service.carml.CarmlEndpoint$$OwbNormalScopeProxy0.doConversion(com/zazuko/service/carml/CarmlEndpoint.java)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)\r\n\tat org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)\r\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)\r\n\tat org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)\r\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)\r\n\tat org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)\r\n\tat org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)\r\n\tat org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)\r\n\tat org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)\r\n\tat org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)\r\n\tat org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)\r\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:681)\r\n\tat org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)\r\n\tat org.apache.meecrowave.cxf.CxfCdiAutoSetup$1.doFilter(CxfCdiAutoSetup.java:122)\r\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)\r\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)\r\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)\r\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)\r\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540)\r\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)\r\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)\r\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)\r\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)\r\n\tat org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382)\r\n\tat org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)\r\n\tat org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895)\r\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732)\r\n\tat org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)\r\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)\r\n\tat org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)\r\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\r\n\tat java.base/java.lang.Thread.run(Thread.java:834)\r\n"}
```

Testfiles:
```
### file  ./example.xml
<characters>
    <character id="0">
        <firstname>Bob</firstname>
        <lastname>XmlSmith</lastname>
        <hair>black</hair>
    </character>
    <character id="1">
        <firstname>Charlie</firstname>
        <hair>orange</hair>
    </character>
</characters>
### file  ./empty-source.json
### file  ./empty-source.xml
### file  ./mapping.carml-xpath.ttl
PREFIX rr: <http://www.w3.org/ns/r2rml#>
PREFIX rml: <http://semweb.mmlab.be/ns/rml#>
PREFIX ql: <http://semweb.mmlab.be/ns/ql#>
PREFIX carml: <http://carml.taxonic.com/carml/>
PREFIX ex: <http://example.org/>
PREFIX schema: <http://schema.org/>

<#person>
        a rr:TriplesMap;

        rml:logicalSource [
                rml:source [
                        a carml:Stream;
                        carml:streamName "stdin"
                ];
                rml:referenceFormulation ql:XPath;
                rml:iterator "/characters/*"
        ];

        rr:subjectMap [
                rr:template "http://example.org/character/{@id}";
                rr:class schema:Person
        ];

        rr:predicateObjectMap [
                rr:predicate schema:givenName;
                rr:objectMap [
                        rml:reference "firstname"
                ]
        ];

        rr:predicateObjectMap [
                rr:predicate schema:lastName;
                rr:objectMap [
                        rml:reference "lastname"
                ]
        ];

        rr:predicateObjectMap [
                rr:predicate ex:hairColor;
                rr:objectMap [
                        rml:reference "hair"
                ]
        ].

### file  ./empty-mapping.ttl
### file  ./example.json
{
  "characters": [
    {
      "id": "0",
      "firstname": "Bob",
      "lastname": "JsonSmith",
      "hair": "black"
    },
    {
      "id": "1",
      "firstname": "Charlie",
      "hair": "orange"
    }
  ]
}

### file  ./mapping.carml.ttl
PREFIX rr: <http://www.w3.org/ns/r2rml#>
PREFIX rml: <http://semweb.mmlab.be/ns/rml#>
PREFIX ql: <http://semweb.mmlab.be/ns/ql#>
PREFIX carml: <http://carml.taxonic.com/carml/>
PREFIX ex: <http://example.org/>
PREFIX schema: <http://schema.org/>

<#person>
        a rr:TriplesMap;

        rml:logicalSource [
                rml:source [
                        a carml:Stream;
                        carml:streamName "stdin"
                ];
                rml:referenceFormulation ql:JSONPath;
                rml:iterator "$.characters[*]"
        ];

        rr:subjectMap [
                rr:template "http://example.org/character/{id}";
                rr:class schema:Person
        ];

        rr:predicateObjectMap [
                rr:predicate schema:givenName;
                rr:objectMap [
                        rml:reference "firstname"
                ]
        ];

        rr:predicateObjectMap [
                rr:predicate schema:lastName;
                rr:objectMap [
                        rml:reference "lastname"
                ]
        ];

        rr:predicateObjectMap [
                rr:predicate ex:hairColor;
                rr:objectMap [
                        rml:reference "hair"
                ]
        ].
```

**Comment by mchlrch** [comment]
I opened a ticket with `carml`: https://github.com/carml/carml/issues/315

**Comment by pmaria** [comment]
@ktk @mchlrch I just released a fix for this in version 0.4.9. Artifacts should be available soon in Maven Central.

**Comment by ktk** [comment]
@pmaria wow world class, thanks Pano!

**Comment by mchlrch** [comment]
Fixed in [v0.0.5](https://github.com/zazuko/carml-service/releases/tag/v0.0.5) by updating the dependencies to carml version 0.4.9
